pybricksdev/connections.py

```diff
@@ -256,22 +256,34 @@ def nus_handler(self, sender, data):
         for line in lines:
             self.line_handler(line)
 
-    def pybricks_service_handler(self, _: int, data: bytes) -> None:
+    def __pybricks_service_handler(self, _: int, data: bytes) -> None:
         if data[0] == Event.STATUS_REPORT:
-            # decode the payload
-            (flags,) = struct.unpack("<I", data[1:])
-            program_running_now = bool(flags & Status.USER_PROGRAM_RUNNING.flag)
-
             # If we are currently downloading a program, we must ignore user
             # program running state changes, otherwise the checksum will be
             # sent to the terminal instead of being handled by the download
             # algorithm
-            if not self.loading:
-                if self.program_running != program_running_now:
-                    logger.info(f"Program running: {program_running_now}")
-                    self.program_running = program_running_now
-                if not program_running_now:
-                    self.user_program_stopped.set()
+            if not self.program_loaded:
```

Comment:
This prevents use of this callback from anything else in the future.

Reply:
I also think there could still be a race condition where the user program starts and stops before the message that the program has finished downloading has been received. You could probably test this by trying to run an empty program.

Reply:
> This prevents use of this callback from anything else in the future.

Good point. Wanted to avoid several nested IF blocks. Will fix that.

> I also think there could still be a race condition where the user program starts and stops before the message that the program has finished downloading has been received. You could probably test this by trying to run an empty program.

You're right. Replicated it by
```diff
diff --git a/pybricksdev/connections.py b/pybricksdev/connections.py
index 2d0dd82..058647e 100644
--- a/pybricksdev/connections.py
+++ b/pybricksdev/connections.py
@@ -390,6 +390,7 @@ class PybricksHub:
                 await self.__send_block(c)
                 pbar.update(len(c))
 
+        await asyncio.sleep(1)
         self.program_loaded = True
```

Reply:
A question regarding `bleak`.

As I understood from you, the events are issued every 500ms. Does that mean that if the program finishes quickly (say in 100ms, the `Status.USER_PROGRAM_RUNNING.flag == True` might not be published at all?

Reply:
> EDIT: from my observation, the events are sent every ~500ms but also on any change. Am I correct?

That is correct.

Reply:
> There is this comment in the code. How come we could get the USER_PROGRAM_RUNNING event prior to the program being uploaded? 

[nus_handler()](https://github.com/pybricks/pybricksdev/blob/3e8e08cdc51d22e15737a027f8447778a2b09277/pybricksdev/connections.py#L225-L257) is another event handler with similar issues that it has to try to handle every possible object state. The checksum is currently sent using the equivalent of MicroPython `print()` function which buffers the output. So it is possible for the user program to start, which triggers a status notification message to be sent, before the last buffered checksum is sent.

Really, we should change the implementation in the firmware to separate the download and run mechanism from the `print()` function, but this is what we have for now.

Reply:
> I've also noticed that when a program finishes quickly and the race condition appears (pybricksdev is not terminated) and I press the Hub's power button, the following sequence of flags is received. 

I had a feeling this would happen. 😄 We probably also need to include a timeout while waiting for the "user program is running" condition. If we don't receive it within one second, we should probably just assume that it was a short running program and we missed the signal.

> Not sure why the POWER_BUTTON_PRESSED is issued 3 times and also the USER_PROGRAM_RUNNING is issued twice. 

You are seeing the effect of messages being sent when an event is triggered and the 500 ms trigger. Since it is possible for messages to be missed due to bad RF conditions, we just keep sending the current state over and over so that when it is finally received, the handling program can do the right thing. This way it is less likely to miss a change in state.

Reply:
> nus_handler() is another event handler with similar issues that it has to try to handle every possible object state. 

I am still a bit lost in the program upload/run/checksum process, to be honest. There is a lot going on and since I have no experience with Bluetooth transfers, all this seems like dark magic 😊. Trying to create some sort of sequence diagram in my head but failing so far.

I guess the missing piece of the puzzle is to find out what the `nus_handler` and `line_handler()` do, and why.